File: backend/app/schemas/audit_entity.py
from datetime import datetime

from pydantic import BaseModel, EmailStr, Field


class AuditEntityBase(BaseModel):
    parent_entity_id: int | None = None
    entity_name: str = Field(..., max_length=255)
    entity_type: str = Field(..., max_length=50)
    entity_class: str = Field(default="company", max_length=50)
    legal_status_id: int | None = None
    legal_status: str | None = Field(default=None, max_length=50)

    registration_no: str | None = Field(default=None, max_length=100)
    tax_identification_no: str | None = Field(default=None, max_length=100)
    contact_person: str | None = Field(default=None, max_length=150)
    contact_email: EmailStr | None = Field(default=None, max_length=150)
    contact_phone: str | None = Field(default=None, max_length=50)
    address: str | None = None
    city: str | None = Field(default=None, max_length=100)
    country: str | None = Field(default=None, max_length=100)
    risk_rating: str | None = Field(default=None, max_length=30)
    is_internal: bool = False
    is_confidential: bool = False
    description: str | None = None
    remarks: str | None = None


class AuditEntityCreate(AuditEntityBase):
    entity_code: str | None = Field(default=None, max_length=50)


class AuditEntityUpdate(BaseModel):
    parent_entity_id: int | None = None
    entity_code: str | None = Field(default=None, max_length=50)
    entity_name: str | None = Field(default=None, max_length=255)
    entity_type: str | None = Field(default=None, max_length=50)
    entity_class: str | None = Field(default=None, max_length=50)
    legal_status_id: int | None = None
    legal_status: str | None = Field(default=None, max_length=50)

    registration_no: str | None = Field(default=None, max_length=100)
    tax_identification_no: str | None = Field(default=None, max_length=100)
    contact_person: str | None = Field(default=None, max_length=150)
    contact_email: EmailStr | None = Field(default=None, max_length=150)
    contact_phone: str | None = Field(default=None, max_length=50)
    address: str | None = None
    city: str | None = Field(default=None, max_length=100)
    country: str | None = Field(default=None, max_length=100)
    risk_rating: str | None = Field(default=None, max_length=30)
    is_internal: bool | None = None
    is_confidential: bool | None = None
    description: str | None = None
    remarks: str | None = None
    is_active: bool | None = None


class AuditEntityResponse(AuditEntityBase):
    id: int
    entity_code: str
    is_active: bool
    created_by: str | None = None
    updated_by: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True,
    }


class AuditEntityListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[AuditEntityResponse]


class AuditEntityMessageResponse(BaseModel):
    message: str
    data: AuditEntityResponse | None = None

File: frontend/src/services/auditEntity.ts
export type AuditEntity = {
  id: number;
  parent_entity_id: number | null;
  entity_code: string;
  entity_name: string;
  entity_type: string;
  entity_class: string;
  legal_status_id: number | null;
  legal_status: string | null;
  registration_no: string | null;
  tax_identification_no: string | null;
  contact_person: string | null;
  contact_email: string | null;
  contact_phone: string | null;
  address: string | null;
  city: string | null;
  country: string | null;
  risk_rating: string | null;
  is_internal: boolean;
  is_confidential: boolean;
  description: string | null;
  remarks: string | null;
  is_active: boolean;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
};

export type AuditEntityListResponse = {
  total: number;
  page: number;
  page_size: number;
  items: AuditEntity[];
};

export type AuditEntityPayload = {
  parent_entity_id?: number | null;
  entity_code?: string | null;
  entity_name: string;
  entity_type: string;
  entity_class: string;
  legal_status_id?: number | null;
  legal_status?: string | null;
  registration_no?: string | null;
  tax_identification_no?: string | null;
  contact_person?: string | null;
  contact_email?: string | null;
  contact_phone?: string | null;
  address?: string | null;
  city?: string | null;
  country?: string | null;
  risk_rating?: string | null;
  is_internal: boolean;
  is_confidential: boolean;
  description?: string | null;
  remarks?: string | null;
};

export type AuditEntityUpdatePayload = Partial<AuditEntityPayload> & {
  is_active?: boolean;
};

type ListParams = {
  page: number;
  pageSize: number;
  search?: string;
  isActive?: boolean;
  entityType?: string;
  entityClass?: string;
  parentEntityId?: number;
  riskRating?: string;
  sortBy?: string;
  sortOrder?: "asc" | "desc";
};

function buildQuery(params: ListParams) {
  const query = new URLSearchParams();

  query.set("page", String(params.page));
  query.set("page_size", String(params.pageSize));
  query.set("sort_by", params.sortBy ?? "id");
  query.set("sort_order", params.sortOrder ?? "asc");

  if (params.search) query.set("search", params.search);

  if (typeof params.isActive === "boolean") {
    query.set("is_active", String(params.isActive));
  }

  if (params.entityType) query.set("entity_type", params.entityType);
  if (params.entityClass) query.set("entity_class", params.entityClass);

  if (params.parentEntityId) {
    query.set("parent_entity_id", String(params.parentEntityId));
  }

  if (params.riskRating) query.set("risk_rating", params.riskRating);

  return query.toString();
}

async function requestJson<T>(
  url: string,
  options?: RequestInit,
): Promise<T> {
  const response = await fetch(url, {
    ...options,
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...(options?.headers ?? {}),
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    const message =
      error?.detail || error?.message || "Audit entity request failed.";

    throw new Error(message);
  }

  return (await response.json()) as T;
}

export async function listAuditEntities(params: ListParams) {
  const query = buildQuery(params);

  return requestJson<AuditEntityListResponse>(
    `/api/backend/audit-entities?${query}`,
    {
      method: "GET",
    },
  );
}

export async function createAuditEntity(payload: AuditEntityPayload) {
  return requestJson<{
    message: string;
    data: AuditEntity;
  }>("/api/backend/audit-entities", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateAuditEntity(
  id: number,
  payload: AuditEntityUpdatePayload,
) {
  return requestJson<{
    message: string;
    data: AuditEntity;
  }>(`/api/backend/audit-entities/${id}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function deactivateAuditEntity(id: number) {
  return requestJson<{
    message: string;
    data: AuditEntity;
  }>(`/api/backend/audit-entities/${id}`, {
    method: "DELETE",
  });
}

export async function restoreAuditEntity(id: number) {
  return requestJson<{
    message: string;
    data: AuditEntity;
  }>(`/api/backend/audit-entities/${id}/restore`, {
    method: "PATCH",
  });
}

export async function permanentDeleteAuditEntity(id: number) {
  return requestJson<{ message: string; data: null }>(
    `/api/backend/audit-entities/${id}/permanent`,
    {
      method: "DELETE",
    },
  );
}
